scheds/rust/scx_bpfland/src/bpf/main.bpf.c

```diff
@@ -554,6 +527,24 @@ static s32 pick_idle_cpu(struct task_struct *p, s32 prev_cpu)
 	idle_smtmask = scx_bpf_get_idle_smtmask();
 	idle_cpumask = scx_bpf_get_idle_cpumask();
 
+	/*
+	 * For tasks that can run only on a single CPU, we can simply verify if
+	 * their only allowed CPU is still usable, online and idle.
+	 *
+	 * Moreover, if local_kthreads is enabled, always dispatch per-CPU
+	 * kthreads directly to their target CPU, independently on its idle
+	 * state.
+	 */
+	if (p->nr_cpus_allowed == 1) {
+		if (bpf_cpumask_test_cpu(prev_cpu, p->cpus_ptr) &&
+		    bpf_cpumask_test_cpu(prev_cpu, online_cpumask) &&
```

Comment:
I wonder if an explicit checking against online_cpumask is indeed necessary. I guess anyway such case will be handled by  scx_bpf_test_and_clear_cpu_idle(prev_cpu). What do you think?

Reply:
I don't think `scx_bpf_test_and_clear_cpu_idle()` is aware of the CPUs that may go offline here, the idle_masks are initialized by `reset_idle_masks()`, that is called only from `scx_ops_enable()`, so IIUC idle masks are not updated on hotplug events.

@htejun should we call `reset_idle_masks()` from `handle_hotplug()`? That would temporarily invalidate the current idle state, but it should converge soon, and in this way offline CPUs would be automatically considered as busy without having to check the online cpumask.

Reply:
It's safe to dispatch to offline CPUs. sched_ext core will just punt them to `SCX_DSQ_GLOBAL` if the target CPU is offline, so the BPF scheduler shouldn't need to worry about that - the next time the task comes around, its `prev_cpu` won't be the offline CPU.

Reply:
With `SCX_DSQ_LOCAL` / `SCX_DSQ_LOCAL_ON` yes, but `dispatch_direct_cpu()` here is actually dispatching to the custom per-CPU DSQ, that can be only consumed by the corresponding CPU. If I dispatch the task and the CPU goes offline, the task won't be consumed.

Reply:
Ah, okay, bpfland implements its own hotplug ops and doesn't restart. Down the line, a better way to deal with this probably is draining the per-cpu DSQ from `ops.cpu_offline()`. The online mask is cleared after the scheduler on that CPU is shut down, so there is still gonna be gaps.

Reply:
Yep, that was my original idea, but I need an API to trigger a re-enqueue of all the tasks in a custom DSQ (I mentioned this during office hours), because I can't use `scx_bpf_consume()` directly from `ops.cpu_offline()`.

Anyway, even if there's a gap and a task is dispatched to a per-cpu DSQ associated to an offline CPU it can still be consumed from other CPUs via `consume_offline_cpus()`. Maybe I can just get rid of these online checks here and just let the tasks be consumed by the other CPUs.